airflow/logging_config.py

```diff
@@ -0,0 +1,71 @@
+# -*- coding: utf-8 -*-
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+import logging
+import os
+import sys
+
+from logging.config import dictConfig
+from airflow import configuration as conf
+
+log = logging.getLogger(__name__)
+
+
+def configure_logging(custom_config_path=None):
+    if custom_config_path is not None:
+        config_path = custom_config_path
+    else:
+        config_path = conf.get('core', 'logging_config_path')
+
+    try:
+        path = os.path.dirname(config_path)
+        sys.path.append(path)
+
+        filename = os.path.basename(config_path).replace('.py', '')
```

Comment:
Does this actually work if it contains '/' in it? Or should this setting be called `loging_config_module`?

i.e. `conf/mylogging.py` vs `conf.mylogging`?

Reply:
The custom config filename can be anything, but it has to be a python file because we expect a python dict, so `conf/mylogging.py` would be fine.

Reply:
Oh right, I see how it's done, because we add the directory of the file to python path.

That might lead to some counter-intuitive behaviour if my config module is "deep" in another module somewhere (say `mycompany.airflow.custom_logging`) which would then add `mycompany/airflow/` into the python search path, meaning if it did it's own imports of `mycompany.x`, they wouldn't work.

At the very least we need to document _what_ we do with the python import path, though my preference here would be to leave any altering of the python path up to the user (via PYTHONPATH env var) and to accept something like what gunicorn does: `$(MODULE_NAME):$(VARIABLE_NAME)` (where VARIABLE_NAME could instead be hard-coded to be LOGGING_CONFIG)

Reply:
Keep in mind that it is appended to the path, not prepended. So it will not alter any existing imports. The only problem might be that the import is already existing.

What would you recommend? Before I searched for a specific file on the PYTHONPATH. But this might be a bit more cumbersome for some users. On @bolkedebruin's suggestion I changed it so you now can set the path to the config file: `logging_config_path=/home/fokko/airflow_config.py`.

Reply:
Oh yes, good point about adding to the end of the path, not the start.

But here's a thought that would make this a smaller change, and mean we don't have to load another file. Given that airflow already loads 'airflow_local_settings.py', we could define a LOGGING_CONFIG dict in airflow.settings, and then use that, rather than a new python file just for logging.

i.e:

- airflow.settings module defines the base/default LOGGING_CONFIG in dict form
- due to the already existing `from airflow_local_settings import *` if that defines a new `LOGGING_CONFIG` it would be used instead.
- `configure_logging()` could then just look at airflow.settings.LOGGING_CONFIG.

Thoughts?

(Forgive me if I missed some discussions about this, I was on vacation)

Reply:
Thank you for the feedback. I looked into how Django is doing this, and learned from here that they just import a module. So in the `airflow.cfg` you'll have to specify:
```
LOGGING_CONFIG_CLASS = my.path.class.LOGGING_CONFIG
```
Which has to be on the PYTHONPATH, but Airflow will just search for this class and import the LOGGING_CONFIG dict, which is then loaded. What do you think about this?

Reply:
Yeah that sounds like a sensible approach to take, (and copying from Django is almost always a good idea to me 😄)